sql/src/main/java/io/crate/operation/reference/information/TablesVersionExpression.java

```diff
@@ -0,0 +1,108 @@
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.reference.information;
+
+import io.crate.Version;
+import io.crate.metadata.RowContextCollectorExpression;
+import io.crate.metadata.table.StoredTable;
+import io.crate.metadata.table.TableInfo;
+import io.crate.operation.reference.RowCollectNestedObjectExpression;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+
+class TablesVersionExpression extends RowCollectNestedObjectExpression<TableInfo> {
+
+    TablesVersionExpression() {
+        addChildImplementations();
+    }
+
+    private void addChildImplementations() {
+        childImplementations.put(Version.Property.CREATED.toString(),
+            new TableVersionCreatedExpression(Version.Property.CREATED));
+        childImplementations.put(Version.Property.UPGRADED.toString(),
+            new TableVersionCreatedExpression(Version.Property.UPGRADED));
+    }
+
+    @Override
+    public Map<String, Object> value() {
+        Map<String, Object> map = super.value();
+        return getValuesMap(map);
+    }
+
+    static class TableVersionCreatedExpression extends RowCollectNestedObjectExpression<TableInfo> {
+
+        TableVersionCreatedExpression(Version.Property property) {
+            addChildImplementations(property);
+        }
+
+        private void addChildImplementations(Version.Property property) {
+            childImplementations.put(Version.CRATEDB_VERSION_KEY.toString(),
+                new TableDetailedVersionExpression(property, Version.CRATEDB_VERSION_KEY));
+            childImplementations.put(Version.ES_VERSION_KEY.toString(),
+                new TableDetailedVersionExpression(property, Version.ES_VERSION_KEY));
+        }
+
+        @Override
+        public Map<String, Object> value() {
+            Map<String, Object> map = super.value();
+            return getValuesMap(map);
+        }
+    }
+
+    static class TableDetailedVersionExpression extends RowContextCollectorExpression<TableInfo, Object> {
+
+        private final Version.Property property;
+        private final String dbKey;
+
+        TableDetailedVersionExpression(Version.Property property, String dbKey) {
+            this.property = property;
+            this.dbKey = dbKey;
+        }
+
+        @Override
+        public Object value() {
+            Version version = null;
+            if (row instanceof StoredTable) {
+                StoredTable storedTable = (StoredTable) row;
+                if (property == Version.Property.CREATED) {
+                    version = storedTable.versionCreated();
+                } else {
+                    version = storedTable.versionUpgraded();
+                }
+            }
+
+            return version == null ? null : Version.toStringMap(version).get(dbKey);
```

Comment:
Wouldn't it be cleaner to have to concrete implementations that know which property of version to access instead of using this verionMap indirection? Would also avoid the map creation